File: Lab Assignment 3/Lab3.py
# Sorting a list of integers
list_int = [4, 65, 2, -31, 0, 99, 2, 83, 782]
list_int.sort()
print(*list_int, "\n")

# Sorting a list of floating point values
list_float = [645.32, 37.40, 76.30, 5.40, -34.23, 1.11, -34.94, 23.37, 635.46, -876.22, 467.73, 62.26]
list_float.sort()
print(*list_float, "\n")

# Sorting a list of characters
list_char = ['z', 'y', 'a', 't', 'p', 'q', 'l', 'k']
list_char.sort()
print(*list_char, "\n")

# Sorting a list of strings
list_string = ["apple", "zebra", "turtle", "yellow", "quest", "buffalo", "elephant"]
list_string.sort()
print(*list_string, "\n")

class Person:
    def __init__ (self, name, age):
        self.name = name
        self.age = age
    def __str__(self):
        return str(self.name) + "-" + str(self.age)

def cmp(x, y):
    return (x > y) - (x < y)

def cmp_age(x, y):
    return cmp(x.age, y.age)

# Sorting a list of objects
list_object = [Person("Hal", 20), 
               Person("Susann", 31),
               Person("Dwight", 19),
               Person("Kassandra", 21), 
			   Person("Lawrence", 25),
               Person("Cindy", 22), 
               Person("Cory", 27), 
               Person("Mac", 19),
               Person("Romana", 27), 
               Person("Doretha", 32),
               Person("Danna", 20), 
               Person("Zara", 23), 
               Person("Rosalyn", 26), 
               Person("Risa", 24), 
               Person("Benny", 28),
               Person("Juan", 33), 
               Person ("Natalie", 25)]
list_object.sort(key = lambda Person: Person.name) # sort by name
# list_object.sort(key = lambda Person: Person.age) # sort by age
print(*list_object, "\n");

File: Lab Assignment 3/Lab3.c
#include <stdio.h>

int type;
/*
	type = 1 (int)
	type = 2 (float)
	type = 3 (char)
	type = 4 (string)
	type = 5 (struct)
*/

typedef struct Person
{
	char* name;
	int age;

} Person;

// Print elements in an array
void printArray(void* arr, int size, int type)
{
	if (type == 1) // int
	{
		for (int i = 0; i < size; i++)
    		printf("%d ", *((int*)arr + i));
	}
	else if (type == 2) // float
	{
		for (int i = 0; i < size; i++)
    		printf("%.2f ", *((float*)arr + i));
	}
	else if (type == 3) // char
	{
		for (int i = 0; i < size; i++)
    		printf("%c ", *((char*)arr + i));
	}
	else if (type == 4) // string
	{	
   		printf("%s", (char*)arr);
	}
	else if (type == 5) // struct
	{
		Person* temp = (Person*)arr;

		for (int i = 0; i < size; i++)
		{
			printf("%s-%d ", temp->name, temp->age);
			temp = ((Person*)temp + 1);

			if (i != size - 1)
				printf(" ");
		}
	}
}

// Compare function for passing in to built-in function 'qsort'
int cmpfunc (const void* a, const void* b)
{
	if (type == 1) // int
	{
		return ((*(int*)a) - (*(int*)b));
	}
	else if (type == 2) // float
	{
		return ((*(float*)a) - (*(float*)b));
	}
	else if (type == 3) // char
	{
		return ((*(char*)a) - (*(char*)b));
	}
	else if (type == 4) // string
	{
		return strcmp((*(char**)a) , (*(char**)b));
	}
	else if (type == 5) // struct
	{
		Person* temp_a = (Person*)a;
		Person* temp_b = (Person*)b;

		// Sort by name first then age
		if(strcmp(temp_a->name, temp_b->name) == 0)
			return (temp_a->age - temp_b->age);

		return strcmp(temp_a->name, temp_b->name);

		/*
		// Sort by age first then name
		if((temp_a->age - temp_b->age) == 0)
			return (strcmp(temp_a->name, temp_b->name);

		return (temp_a->age - temp_b->age);
		*/
	}

	if(type != 5)
		printf("\n\n");
}

int main()
{
	int arr_size;

	printf("\n");

	// Sorting array of integers
	type = 1;
	int arr_int[] = {4, 65, 2, -31, 0, 99, 2, 83, 782};
	arr_size = sizeof(arr_int) / sizeof(arr_int[0]);
	qsort(arr_int, arr_size, sizeof(int), cmpfunc);
	printArray(arr_int, arr_size, type);
	printf("\n\n");

	// Sorting array of floating point values
	type = 2;
	float arr_float[] = {645.32, 37.40, 76.30, 5.40, -34.23, 1.11, -34.94, 23.37, 635.46, -876.22, 467.73, 62.26};
	arr_size = sizeof(arr_float) / sizeof(arr_float[0]);
	qsort(arr_float, arr_size, sizeof(float), cmpfunc);
	printArray(arr_float, arr_size, type);
	printf("\n\n");

	// Sorting array of characters
	type = 3;
	char arr_char[] = {'z', 'y', 'a', 't', 'p', 'q', 'l', 'k'};
	arr_size = sizeof(arr_char) / sizeof(arr_char[0]);
	qsort(arr_char, arr_size, sizeof(char), cmpfunc);
	printArray(arr_char, arr_size, type);
	printf("\n\n");

	// Sorting array of strings
	type = 4;
	char* arr_string[] = {"apple", "zebra", "turtle", "yellow", "quest", "buffalo", "elephant"};
	arr_size = sizeof(arr_string) / sizeof(arr_string[0]);
	qsort(arr_string, arr_size, sizeof(char*), cmpfunc);
	for (int i = 0; i < arr_size; i++)
	{
		int string_length = strlen(arr_string[i]) + 1;
		printArray(arr_string[i], string_length, type);
		
		if(i != arr_size-1)
			printf(" ");
	}
	printf("\n\n");

	// Sorting array of structures
	type = 5;
	Person arr_struct[] = {[0].name = "Hal", [0].age = 20,
						   [1].name = "Susann",[1].age = 31,
						   [2].name = "Dwight",[2].age = 19,
						   [3].name = "Kassandra",[3].age = 21,
						   [4].name = "Lawrence",[4].age = 25,
						   [5].name = "Cindy",[5].age = 22,
						   [6].name = "Cory",[6].age = 27,
						   [7].name = "Mac",[7].age = 19,
						   [8].name = "Romana",[8].age = 27,
						   [9].name = "Doretha",[9].age = 32,
						   [10].name = "Danna",[10].age = 20,
						   [11].name = "Zara",[11].age = 23,
						   [12].name = "Rosalyn",[12].age = 26,
						   [13].name = "Risa",[13].age = 24,
						   [14].name = "Benny",[14].age = 28,
						   [15].name = "Juan",[15].age = 33,
						   [16].name = "Natalie",[16].age = 25};
	arr_size = sizeof(arr_struct) / sizeof(arr_struct[0]);
	qsort(arr_struct, arr_size, sizeof(Person), cmpfunc);
	printArray(arr_struct, arr_size, type);
	printf("\n");

    return 0;
}